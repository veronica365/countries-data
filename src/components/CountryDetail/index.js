import PropTypes from 'prop-types';

export default function CountryDetail({ country }) {
  return (
    <article className="country detail">
      <div className="country-info">
        <h2 className="country-name">{country.name}</h2>
        <span className="country-flag">{country.emoji}</span>
        <h3 className="country-name">
          <i className="fa fa-building" />
          {' '}
&nbsp;&nbsp;
          {country.capital}
        </h3>
        <div className="country-data">
          <span className="country-data-value">
            <i className="fa fa-mobile" />
            {country.phone_code}
          </span>
          <span className="country-data-value">
            <i className="fa fa-money" />
            {country.currency}
            {' '}
            (
            {country.currency_symbol}
            {' '}
            )
          </span>
          <span className="country-data-value">
            <i className="fa fa-globe" />
            {country.region}
            ,
            {country.subregion}
          </span>
          <span className="country-data-value">
            <i className="fa fa-map-marker" />
            {parseFloat(country.latitude).toFixed(2)}
            {'  '}
            {'  '}
            {parseFloat(country.longitude).toFixed(2)}
          </span>
          <span className="country-data-value">
            <i className="fa fa-clock-o" />
            Timezones
          </span>
        </div>
        <div className="country-data">
          <h4>
            All time zones in
            {country.name}
            :
          </h4>
        </div>
        <div className="country-data">
          <div className="country-data-value">
            {country.timezones?.map((timezone, idx) => (
              <p className="timezone" key={timezone.zoneName}>
                {idx + 1}
                .
                {timezone.zoneName}
                {' '}
                {timezone.gmtOffsetName}
                ,
                {' '}
                {timezone.abbreviation}
                {' '}
                (
                {timezone.tzName}
                ),
              </p>
            ))}
          </div>
        </div>
      </div>
    </article>
  );
}

CountryDetail.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  country: PropTypes.object.isRequired,
  //   country: PropTypes.shape({
  //     id: PropTypes.number.isRequired,
  //     name: PropTypes.string.isRequired,
  //     phone_code: PropTypes.string.isRequired,
  //     currency: PropTypes.string.isRequired,
  //     latitude: PropTypes.string.isRequired,
  //     longitude: PropTypes.string.isRequired,
  //   }).isRequired,
};
