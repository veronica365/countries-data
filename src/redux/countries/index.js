import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';
import { API } from '../../constants';

const findItem = (id, countries) => countries.filter((country) => String(country.id) === id);
const searchItem = (param, countries) => countries.filter(
  (country) => String(country.id) === param
      || String(country.name).toLowerCase().includes(String(param).toLowerCase())
      || String(country.currency).toLowerCase().includes(String(param).toLowerCase())
      || String(country.capital).toLowerCase().includes(String(param).toLowerCase())
      || String(country.region).toLowerCase().includes(String(param).toLowerCase())
      || String(country.currency_name).toLowerCase().includes(String(param).toLowerCase())
      || String(country.subregion).toLowerCase().includes(String(param).toLowerCase()),
);

export const getData = createAsyncThunk('get', async (_, thunkAPI) => {
  try {
    const { countries } = thunkAPI.getState().countries;
    if (countries.length) return countries;
    const { data } = await axios.get(`${API}`);
    return data;
  } catch (error) {
    return thunkAPI.rejectWithValue('Something went wrong');
  }
});
export const findCountry = createAsyncThunk('find', async (id, thunkAPI) => {
  try {
    const { countries } = thunkAPI.getState().countries;
    if (countries.length) return findItem(id, countries);
    const { data } = await axios.get(`${API}`);
    return findItem(id, data);
  } catch (error) {
    return thunkAPI.rejectWithValue('Something went wrong');
  }
});
export const searchCountry = createAsyncThunk(
  'search',
  async (param, thunkAPI) => {
    try {
      const { countries } = thunkAPI.getState().countries;
      if (countries.length) return searchItem(param, countries);
      const { data } = await axios.get(`${API}`);
      return searchItem(param, data);
    } catch (error) {
      return thunkAPI.rejectWithValue('Something went wrong');
    }
  },
);

const countrieslice = createSlice({
  name: 'countries',
  initialState: {
    countries: [],
    countriesBySearch: [],
    countryById: {},
    isLoading: false,
    error: null,
  },
  reducers: {
    setCountryById: (state, action) => ({
      ...state,
      countryById: action.payload,
      countriesBySearch: [],
      isLoading: false,
    }),
  },
  extraReducers: (builder) => {
    builder.addCase(getData.pending, (state) => ({
      ...state,
      isLoading: true,
    }));
    builder.addCase(getData.fulfilled, (state, action) => ({
      ...state,
      countries: action.payload,
      isLoading: false,
    }));
    builder.addCase(getData.rejected, (state, action) => ({
      ...state,
      error: action.payload,
      isLoading: false,
    }));
    builder.addCase(findCountry.pending, (state) => ({
      ...state,
      isLoading: true,
    }));
    builder.addCase(findCountry.fulfilled, (state, action) => ({
      ...state,
      countryById: action.payload[0] || {},
      isLoading: false,
    }));
    builder.addCase(findCountry.rejected, (state, action) => ({
      ...state,
      error: action.payload,
      isLoading: false,
    }));
    builder.addCase(searchCountry.pending, (state) => ({
      ...state,
      isLoading: true,
    }));
    builder.addCase(searchCountry.fulfilled, (state, action) => ({
      ...state,
      countriesBySearch: action.payload,
      isLoading: false,
    }));
    builder.addCase(searchCountry.rejected, (state, action) => ({
      ...state,
      error: action.payload,
      isLoading: false,
    }));
  },
});

export const { setCountryById } = countrieslice.actions;

export default countrieslice.reducer;
