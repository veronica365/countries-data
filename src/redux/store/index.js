import { configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import countryReducer from '../countries';

const store = configureStore({
  reducer: {
    countries: countryReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
  devTools: process.env.NODE_ENV !== 'production',
});

export default store;
