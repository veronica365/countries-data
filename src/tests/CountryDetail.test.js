import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import CountryDetail from '../components/CountryDetail';
import { muchCountryData } from './__Mocks__/countries';
import store from '../redux/store';

describe('CountryDetail', () => {
  it('should display all the country details', () => {
    const { container } = render(
      <Provider store={store}>
        <CountryDetail country={muchCountryData} />
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
