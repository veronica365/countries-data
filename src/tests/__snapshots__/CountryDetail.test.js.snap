// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryDetail should display all the country details 1`] = `
<div>
  <article
    class="country detail"
  >
    <div
      class="country-info"
    >
      <h2
        class="country-name"
      >
        Afghanistan
      </h2>
      <span
        class="country-flag"
      >
        ðŸ‡¦ðŸ‡«
      </span>
      <h3
        class="country-name"
      >
        <i
          class="fa fa-building"
        />
         
        Â Â 
        Kabul
      </h3>
      <div
        class="country-data"
      >
        <span
          class="country-data-value"
        >
          <i
            class="fa fa-mobile"
          />
          93
        </span>
        <span
          class="country-data-value"
        >
          <i
            class="fa fa-money"
          />
          AFN
           
          (
          Ø‹
           
          )
        </span>
        <span
          class="country-data-value"
        >
          <i
            class="fa fa-globe"
          />
          Asia
          ,
          Southern Asia
        </span>
        <span
          class="country-data-value"
        >
          <i
            class="fa fa-map-marker"
          />
          33.00
            
            
          65.00
        </span>
        <span
          class="country-data-value"
        >
          <i
            class="fa fa-clock-o"
          />
          Timezones:
           
          <b>
            1
          </b>
        </span>
      </div>
      <div
        class="country-data"
      >
        <h4>
          All time zones in
          Afghanistan
          :
        </h4>
      </div>
      <div
        class="country-data"
      >
        <div
          class="country-data-value"
        >
          <p
            class="timezone"
          >
            1
            .
            Asia/Kabul
             
            UTC+04:30
            ,
             
            AFT
             
            (
            Afghanistan Time
            ),
          </p>
        </div>
      </div>
    </div>
  </article>
</div>
`;
